File: java/Heap_sort.java
//Minheap 내림차순정렬
public class Heap_sort{
	
	int[] arr;
	public Heap_sort(int[] arr){
		this.arr=arr;
	}
		
	void buildHeap(int n){
		for(int i=n/2;i>0;i--)
			heapify(i,n);
	}

	void heapify(int k,int n){
		int left=2*k,right=2*k+1;
		int smaller;

		if(right<=n) // exist both childNode based on k parentNode 
			smaller=(arr[left]>arr[right] ? right : left);
		else if(left<=n) // exist left childNode
			smaller=left;
		else
			return ;

		if(arr[smaller]<arr[k]){
			int temp=arr[smaller];
			arr[smaller]=arr[k];
			arr[k]=temp;

			heapify(smaller,n);
		}
	}

	void sort(int n){
		buildHeap(n);

		for(int i=n;i>=2; i--){
			int temp=arr[1];
			arr[1]=arr[i];
			arr[i]=temp;

			heapify(1,i-1);
		}
	}

	public static void main(String[] args){
		int[] unsorted={0,1,7,8,3,9,4,2,6};

		Heap_sort h=new Heap_sort(unsorted);
		h.sort(unsorted.length-1);

		for(int i=unsorted.length-1; i>=1; i--)
			System.out.printf("%d ",unsorted[i]);
		System.out.println();

	}
}

File: python3/heapSort.py
#minheap && descending sort
def buildHeap(arr):
    n=len(arr)-1 #approch index (0~len(arr)-1)
    for i in range(n//2,-1,-1):
        heapify(arr,i,n)
        

def heapify(unsorted,key,size):
    smaller=key
    left=2*key+1; right=2*key+2

    if(left<=size and unsorted[left]<unsorted[smaller]):
        smaller=left
    if(right<=size and unsorted[right]<unsorted[smaller]):
        smaller=right

    if(smaller!=key):
        unsorted[smaller],unsorted[key]=unsorted[key],unsorted[smaller]
        heapify(unsorted,smaller,size)
        
def heapSort(arr):
    buildHeap(arr)
    
    for i in range(len(arr)-1,0,-1):
        arr[0],arr[i]=arr[i],arr[0] #temp arr[0] arr[-1]
        heapify(arr,0,i-1)


arr=list(int(input()) for i in range(int(input())))
heapSort(arr)
print(*arr[::-1],sep='\n')

